app/controllers/setup_controller.rb

```diff
@@ -93,4 +94,15 @@ def minion_not_found(exception)
       format.json { render json: exception.message, status: :not_found }
     end
   end
+
+  def check_empty_settings
+    return unless settings_params.values.any?(&:empty?)
```

Comment:
Yes, I agree. With form objects we will be able to mark wrong fields in an independient fashion, however this will be tricky because all of them are mapped to the `Pillar` model in reality, whose `value` semantics and validations highly depend on the `pillar` key (`country` can only accept one value from a list, an `email` needs to pass a regexp...). Maybe we could add some facade models that actually use `Pillar` as the real backend but they are only for making it easier to use form objects...